### DISCUSSION TITLE
Enhance the comment functionality of Workflow

### DISCUSSION THREAD

**Discussion by dalbard** [umbraco/Umbraco.Workflow.Issues#27] [Ideas]
We have a valuable feature request that could greatly enhance Umbraco Workflow. Our inspiration for this request comes from the comments functionality found in PDF and Word documents. The goal of this feature is to eliminate the need for multiple file copies in emails and chats, while simultaneously providing tighter control over GDPR protected information. Here is how we envision the basic use-case/user story for this feature:

When a user sends an item for review via Workflow, the recipient should receive a link that directs them to a preview of the item. From there, the reviewer can highlight a specific part of the item and attach a comment to it, similar to the functionality found in PDF and Word documents.

All participants in the review process should be able to view each other's comments and reply to them. Essentially, the editor should have the ability to make changes to the item based on the comments, and once all comments have been addressed, the editor can mark the item as corrected/completed.

It should be possible for the person responsible for reviewing the item (highlighting and commenting) to do so without being a user in the Umbraco installation. It should be sufficient to simply email the preview link of the item to the designated reviewer.

The item being reviewed can include different types of pages on the website, as well as various assets in the backoffice, such as PDF-generated content.

We believe this feature is highly valuable as it eliminates the need to send different versions of files through emails or chats. Additionally, the ability to highlight specific parts of a page/document and attach comments allows for more detailed and accurate feedback. This would establish a centralized "one-point-of-truth" for the item under review. Furthermore, this feature enhances the control of GDPR protected information by ensuring that information is only processed within Umbraco and not in other places like emails.

**Comment by nathanwoulfe** [comment]
Hi @dalbard, I'll need to take some time to digest this and consider the potential implications. Couple of initial points to note:

- Workflow can not work with Media, as there is no draft state 
- Workflow is chronological, content is passed from group to group, but can be sent back to previous groups. Not sure that allowing the editor to make changes at any time would fit in that paradigm
- Previewing content requires an active Backoffice user session
- How would non-Umbraco users provide comments? How would these be attributed? How would Workflow know who to email if no user exists?
- Comments can be added at each stage of the workflow by the approving user/s, those comments can reference page specifics if required

Sending file versions via email/chat is not a problem Workflow aims to solve - Workflow provides tooling for content approvals within the CMS. If business processes require distributing files outside Umbraco, that's not something I see Workflow aiming to replace. In my experience, reviewing content within the CMS is more valuable than reviewing same as offline documents, as the latter lacks the context that comes with the webpage.

**Comment by InfobricGroup** [comment]
Hi @nathanwoulfe, Thanks for the feedback and thoughts.

Let me rephrase our request since we now have more understanding or your point of view.

The main point here is that we want to utilize the main functionality of workflow for how to send out content and approval process, since you already have this main process & logic engine in place. But what we feels is missing is the step before approval process. The "content review/feedback" stage, where our subject experts / local market experts can leave feedback/annotations on the content during the writing process, without the need of knowing how to work in the backoffice.

Example of how ProjectHuddle have done it
![image](https://github.com/umbraco/Umbraco.Workflow.Issues/assets/86775159/7fd6dae0-c82a-4408-a4d2-7c86d93670f6)

And how Filestage have done it
![image](https://github.com/umbraco/Umbraco.Workflow.Issues/assets/86775159/2197e1e5-8bca-40b9-ae6f-bc7dd3784a29)


We are extending our backoffice to take more than just webpages, we will also have a PDF "generator" that is creating pages but we are presenting as PDF documents so they would be in Umbraco world be classified as page (since it is going to be a HTML page on the website and as a perfect designed downloadable PDF document) and if that the limiting factor then so be it, we will not be able to send media documents if they are not part of the mage - even though i don't understand what "media attachments" in the content approval is then. And as we have understood your "offline approval" feature that state "secure review link is provided and can be approved without the need to about or log in to the backoffice" meant that one didn't need to be an actual user in the backoffice. How ever if that is a must - so be it, we would then only create a user role that more or less only have access to this then, so that is not a problem, but anyhow, I degrees now back to the point.

Since we are building up a new platform for our large enterprise website with pages, e-commerce products, whitepaper, articles, press releases, customer stories and more in an international perspective (with teams in multiple countries) and we want more active feedback from our colleagues, especially those that are subject experts or knows the local market best during the writing process in a more collaborative way. Therefore, a review flow that those colleagues that the individual editor is selecting to get feedback from can get an email with a link so they can preview the page and mark on the page and add a comment/annotation with what they want to change/redo. After that the editor can answer the comment/annotation if they need more information or if they have made the necessary changes and "check-off" the different comments as done/dismissed or what the person that send it in to sign off on the fix and when all is finished then send it through the approval process.

In this way, the editor does not need to write the page in word/design software and send that as a copy through email to all the internal stakeholders and then need to remind everyone about what version of the document that is the latest. It the editor can create the page in Umbraco and send it for review to our internal stakeholders in a collaborative way and they can directly give the editor feedback (including the X & Y coordinates / marking on the object/text) then the editor can focus on the right stuff by creating value-adding content and be more effectively since they do not need to rewrite the finish word/design file document as a page in Umbraco and instead can just send it for publish approval.

And since we are using customer case/stories as part of our page and want internal feedback on those too we would subsequently minimize how that parties GDPR protected information is spread and we have a one-point of truth for it and all the versions of the page.

But this maby is more of an idea of a new collaborate annotation tool feature - all I want is to find a effificent way of working for all of our editor - lets find a god solution togther (or you can dismiss all this and say "damn, what a bad idea :P ") :)

**Comment by nathanwoulfe** [comment]
Hi @InfobricGroup, thanks for the additional details.

I can see the value in the functionality as described, but still feel it is a very different concept to what Workflow offers - this is very much about collaborative content review outside the CMS (or at least not requiring CMS users), while Workflow is a linear review for content approvals within the CMS.

There could be potential for - as you described - a collaborative annotations feature. Would definitely come with some technical challenges (the ProjectHuddle and Filestage examples are entire companies dedicated to delivering just this tool, so there's obviously substantial complexity in the products).

I'm certainly not saying this won't be considered, but will need to take some time to properly consider how it would fit into Workflow, and if it's something we can fully deliver.

**Comment by InfobricGroup** [comment]
Hi @nathanwoulfe ,

Thank you for your feedback. And I fully understand - If not within Workflow then you guys have got a tips for a new product feature :)

All the best.

**Comment by nathanwoulfe** [comment]
@dalbard, @InfobricGroup keep an eye on this though - I've converted the issue to a discussion so we can keep it around for a while, and see if anyone else shares ideas.
